### DISCUSSION TITLE
Am I  missing a using for AddFileLogger?

### DISCUSSION THREAD

**Discussion by nick5454** [cake-build/cake#3206] [Q&A]
`		
using Cake.Common.Tools.DotNetCore.MSBuild;
var dotNetBuildSettings = new DotNetCoreBuildSettings
		{
			Configuration = BuildParameters.Configuration,
			NoRestore = true
		};

		dotNetBuildSettings.AddFileLogger(new MSBuildFileLoggerSettings { LogFile = BuildParameters.LogPath });
		DotNetCoreBuild(BuildParameters.SolutionFile.ToString(), dotNetBuildSettings);	`

My addings:
`#addin "nuget:https://api.nuget.org/v3/index.json?package=Cake.Incubator&version=5.1.0"
#addin "nuget:https://api.nuget.org/v3/index.json?package=Cake.Npm&version=0.17.0"
#addin "nuget:https://pkgs.dev.azure.com/***/nuget/v3/index.json?package=Cake.DotNetCoreEFAzure"
#addin "nuget:https://api.nuget.org/v3/index.json?package=Cake.Plist"
#addin "nuget:https://api.nuget.org/v3/index.json?package=Cake.FileHelpers&version=3.3.0"
#addin "nuget:https://api.nuget.org/v3/index.json?package=Cake.AppCenter"

#addin nuget:?package=Cake.Git&version=0.22.0

#tool dotnet:?package=GitVersion.Tool&version=5.1.2
#tool dotnet:?package=SignClient&version=1.2.109
#tool dotnet:?package=GitReleaseManager.Tool&version=0.11.0

#addin "Cake.Issues"
#addin "Cake.Issues.MsBuild"
#addin "Cake.Issues.Reporting"
#addin "Cake.Issues.Reporting.Sarif"

#tool "nuget:?package=MSBuild.Extension.Pack"`

And the error "'DotNetCoreBuildSettings' does not contain a definition for 'AddFileLogger' and the best extension method overload 'DotNetCoreMSBuildSettingsExtensions.AddFileLogger(DotNetCoreMSBuildSettings, MSBuildFileLoggerSettings)' requires a receiver of type 'DotNetCoreMSBuildSettings'
"

I'm sure this is an easy one, but i'm missing it. Can anyone help?

**Comment by AdmiringWorm** [comment]
I don't see that function being available for [`DotNetCoreBuildSettings`](https://cakebuild.net/api/Cake.Common.Tools.DotNetCore.Build/DotNetCoreBuildSettings/), however it seems to be available for [`DotNetCoreMSBuildSettings`](https://cakebuild.net/api/Cake.Common.Tools.DotNetCore.MSBuild/DotNetCoreMSBuildSettings/).

So something similar to:
```cs
var dotNetBuildSettings = new DotNetCoreBuildSettings
{
    Configuration = BuildParameters.Configuration,
    NoRestore = true,
    MSBuildSettings = new DotNetCoreMSBuildSettings()
        .AddFileLogger(new MSBuildFileLoggerSettings { LogFile = BuildParameters.LogPath })
};
```

Should work (or so it seems).

**Reply by nick5454** [reply]
That worked, but the log file wasn't generated correctly,







That works, but I don't get xml logging for test results and I didn't see a way to specify a XmlFileLogger? Is this possible or not yet?

```
.WithLogger(
					Context.Tools.Resolve("MSBuild.ExtensionPack.Loggers.dll").FullPath,
					"XmlFileLogger",
					$"logfile=\"{BuildParameters.LogPath}\";verbosity=Detailed;encoding=UTF-8"
					)
				);
```

**Reply by AdmiringWorm** [reply]
There isn't anything out-of-the-box method to add an XML logger (outside of using `WithLogger`), unfortunately.

> but I don't get xml logging for test results

Do you mean to get the same logging functionality when running `dotnet test` as you would with `dotnet build`?
If so, then no that isn't possible, even when running the executable yourself (AFAIK).

**Reply by nick5454** [reply]
Here is my non core build

```		
        MSBuild(BuildParameters.SolutionFile.ToString(), configurator =>
			configurator
				.SetConfiguration(BuildParameters.Configuration)
				.WithProperty("Platform", BuildParameters.Platform)
				.WithProperty("BuildIpa", "true")
				.WithProperty("CodesignKey",  BuildParameters.CertTypes[BranchProject].IOSTeam)
				.WithProperty("CodesignProvision", provisionType) 
				.WithProperty("MtouchNoSymbolStrip", "false")
				.WithProperty("ArchiveOnBuild", "true")
				.WithLogger(
					Context.Tools.Resolve("MSBuild.ExtensionPack.Loggers.dll").FullPath,
					"XmlFileLogger",
					$"logfile=\"{BuildParameters.LogPath}\";verbosity=Detailed;encoding=UTF-8"
					)
				);```

**Reply by nick5454** [reply]
I saw last week I may be able to add it, but I didn't see a way to specify a custom logger (XmlFileLogger). I need these results in xml for the Serif build output.

**Comment by pascalberger** [comment]
If you only need to to pass to Cake.Issues using a binary log might be an easier solution:

```
var msBuildSettings = new DotNetCoreMSBuildSettings();
msBuildSettings.WithLogger(
    "BinaryLogger," + context.Tools.Resolve("Cake.Issues.MsBuild*/**/StructuredLogger.dll"),
    "",
    BuildParameters.LogPath);

DotNetCoreBuild(BuildParameters.SolutionFile, new DotNetCoreBuildSettings
{
    Configuration = BuildParameters.Configuration,
    NoRestore = true,
    MSBuildSettings = msBuildSettings
});
```

The above example uses the `StructuredLogger.dll` assembly shipped with Cake.Issues.MsBuild and makes sure that the binary log is written in a format supported by Cake.Issues.MsBuild, and not dependend on installed Visual Studio version.

**Reply by nick5454** [reply]
Thanks! I will try that

**Reply by pleonex** [reply]
Another way to get the path of `StructuredLogger.dll` is from the `BinaryLog` type:
```csharp
string binaryLoggerAssembly = typeof(Microsoft.Build.Logging.StructuredLogger.BinaryLog).Assembly.Location;
```
It should get the one from the output directory, that Cake.Issues.MsBuild will use.

It doesn't involve the tool resolution, which simplify the usage for Cake.Frosting projects (you don't need to manually install "Cake.Issues.MsBuild" as a tool via `InstallTool` but just as a regular ProjectReference).
